### PR TITLE
john: 1.9.0-Jumbo-1 -> rolling-2404

### LINKED ISSUE #267978: John - ssh2john: AttributeError: module 'base64' has no attribute 'decodestring'
### Describe the bug
ssh2john produce error while trying to process ssh key 

### Steps To Reproduce
1. Install john in any way - `nix-shell -p john`
2. Generate ssh key with password - `ssh-keygen  -f ./id_test`
3. run - `ssh2john.py ./id_test` 

### Expected behavior
KALI:
```
$ ssh-keygen  -f ./id_rsa
Generating public/private rsa key pair.
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in ./id_rsa
Your public key has been saved in ./id_rsa.pub
The key fingerprint is:
SHA256:YSodrYxUJ1T1O9iK5Gtg3Ry4QWCDl0ZD/gM6oi/8JWA casual@kali
The key's randomart image is:
+---[RSA 3072]----+
|     =@+o..      |
|    .++*.  .     |
|    .o+.+.  .    |
|   . = Bo..o .   |
| E. = =.S=..+    |
|.... oooooo. .   |
|o  . o .o .      |
|.o  o   ..       |
| .o.   ..        |
+----[SHA256]-----+

kali% ssh2john  ./id_rsa
./id_rsa:$sshng$6$16$bfb281aa033491761246051e98e0cdea$1894$6f70656e7373682d6b65792d7631000000000a6165733235362d637472000000066263727970740000001800000010bfb281aa033491761246051e98e0cdea000000180000000100000197000000077373682d727361000000030100010000018100ba09be022e5d14ec5d787bd3ddd6495674de42ecde262418db1906558e6a0d1973f1996ff51ab1e2310462d56ba6bff59044cbf5311c6a03be1c6a1d78880ed879c40d0aa2bbe466084d01d197aad53908e4d03044bc4797eed6f8f880c583fe6ec5159f91662f237772828162fed881398b73fb22f473f3e1850bbde3180e6fe1659f0882eeeb243de29e08ec49ebdb63c571de7e4950c1bdfe70f32e0c2aa307ea3ab3f705bb42d1fc63695a20437fa615ddc2a44451de17d5ce81943b583efc29459fc7bec3081161dec237660ab9f9621b4b6c2e151adfefe6d1da94d7214e3c27f63b2a8cf1ed7c8d4b2b85b210fcadba406ff4b8407292e2151b0546fee2d1c090c6aa806eaab9478df9f69fa94df31a87bcdc1d94caedca2dc5a18a538927b4e78e4f0f0ea3c3a79a0bc2a6c2dcbfc4bdc6c1669035ea10b81505fe69d54f4a21b922c02a67cc571ef5ea90e7763b79efbfe59ae7220e4e22d7bfd1442c561f3ec16dab0df6c3f9b375248aa800a10140cba4cb609fed2a68fc9ae88f000005805cb2a240fa5d0f4c8f15e25fee453341f29f9a50ebe2eb2399462bbc0b45f505081f52c11be73a65f3daf466a166d43ac4e02d1db7ea10a66e943d9378be50fc7c06c061db6f46852f4e291fe4fbf59c7540f09e86c810570ec8ff674d3c4da8f88cfa35a65cacf80d9f475d0d79a28d62e355f6bdbe03db19fdb766dcb5363f34ae886e786cf4275cc91c3139e6b17516f03ffc32c90698bdd9c9ec162d37bace397206f5dac7299281386b142ea790e434bf9fc0e6da33553bd93e6ebb33539933689303704027b0e6eeecbac1fcb2290a7c2582b295504d083e8fc370dd32326ae5e08a7d458085cca2853f0c283532bde49ff3e455ce0bc1f0faca6f48db04d293a50d3502062011881393f57e93d1573119ca301bd6e101c6ad61bfa4ff5d644e4091de82f38824b747c24c938edb828faa1c474d507a5234482d4e6f06672f1f9180415796d842c5c4eeb08f548dafc661e9b57b544783fc9113e964411e309743c140069ce32de817377d7436d3c9bf296d8200bd83d2075888b8197dd2cda5718683beb0c05f10ee8a4919bbc71e9061575e25551be7fb50fbc6643a5dc11d069f1038b0d198a41c402d2f822dc095a32c5898c94a89dfe9388b3449fc58b3755678519877527cc0aa35a4e859278dc1e979e2fcebc10e08a47f0480422986cc1b9012e15db319a728473f278f01a01155c22e42223e1e7c95d96a4bb9d7bdab26090f71e54bc7c917e47ba0c03e0489ec0e790a7a60bb4673970b5e868f82bdd4484aaff2e42caee23e6145ca1c6c2c103e1dc8418a25727e8b3b9471a052bb342a3b48056536d770f342329bb1aa7aa27502fbbe42e67a997087f4f8023e53b7037ae30b774fe867d59aefd90c94f071a43fcb4342cd7b4621bb0cb4340df0a7cc39e3017c8f169070ed9d26ab378b26bc24c7f70c6874b11d940e56eb5b7e0301b098ccc7025312188e60e1887e3d91890c1d234a9ecbc075ec067370e419d7c947e0339c594a9a261e32a8963bce7bcb401bbdb54fdfafa4b7354c346a888afcbd2dba4c681fd31fa70052fc61462e4b9e8096c34913d6f745989ce1c34a91eeff22ebf05fd30c0664c996e975ff4bbc9d1b553f2f2933103503bdd8e7f1991ddd5d54b88516f0e3529170e2b91db1a20377e03ec83fb7486c4a9ee1897cb94c6911774ad6b8ce7b028ddc6146eeab23d74ca194e64b9d6cab3b8c71659edb97512476f99792c6874b9c66c23de9e036b107c1c8ff0bf82afcf28a769ec1e2b18df09f17c6fec237ce19b21cf5c7d60c2e9647e05df53faa2088c5adb0ea9dd495e8362c7e96cfb69c0b1d87b0dd75ae61dfc7f8ee4460e8a635adddcb0ff2c45da66f354f29925cf3c94bacbed8947e75bced86388c848b88e29e25cd7ce2d6159b6d2bacf25b30e24a41eeb87c39d9f7fcbab01996ca299421c73f016403d350465c5035fe3a9e72d11b7cee5c494d38278aaabc3b496bf6b742a4d4c19690ba4d81949c5b2f44ea88b9b8127c33b3b42c47f79d3cba5dc74c417f0f12805882b19f2481a85bf53c80a58889aabe4fe75dc3fc4e02e0054c3daf2281d9f101c3301671fcd356e4227977858fee5d8febd49fef5db795b173a3e5d781d15840ba14806c2396f999aad9e4af5ef04a9dbab749a5710d063a36fc0e174ad2656a39ebf1b139006cd560bc4c6fbfed2a576232c510de19519e19ea8214542046eac23819922523e1c1d0a6d9d0f3335fdcf49a94077a9561e68c75b09c11db4e198cb781da9225f3552aaba34ade79c5f47166fc53678b7f1861ba3bf444bfd8d4e788ed93a572318eb5926c3398e3d01c8404a4e8f3ba5f7d6f28fcd9f1edc1aad4571351979c29d9ed9ca35b7f6a26879074e4ba7fb45855c9107d64c6add526294f97fccffb548f93963180adcb59779ddb41af4c655a9a1128292d899b5c8b734ad943227301ee1b51218c51f29dd4bc973615d15e0dc2b74f$24$486

```

### Screenshots
```
tmp ➤ ssh-keygen  -f ./id_rsa -t rsa
Generating public/private rsa key pair.
./id_rsa already exists.
Overwrite (y/n)? y
Enter passphrase (empty for no passphrase):
Enter same passphrase again:
Your identification has been saved in ./id_rsa
Your public key has been saved in ./id_rsa.pub
The key fingerprint is:
SHA256:Cj13/mPjbTBnT0zKOA94Otz8MQNkPxiEZO8TgcCoaiU casual@Casual-PC
The key's randomart image is:
+---[RSA 3072]----+
|       o.oooo    |
|      . ..oo .   |
|     .      *    |
|  E o.     + =  .|
|   +. o S ..=oo+ |
|  o  . + o. Bo=.o|
| .    .  ..= O+o |
|          +.*.o+.|
|           +o=o  |
+----[SHA256]-----+
tmp ➤ ssh2john.py  ./id_rsa
Traceback (most recent call last):
  File "/nix/store/vprysv8g6xxzn5y56ndhqz3dk77l5inn-john-1.9.0-jumbo-1/bin/.ssh2john.py-wrapped", line 194, in <module>
    read_private_key(filename)
  File "/nix/store/vprysv8g6xxzn5y56ndhqz3dk77l5inn-john-1.9.0-jumbo-1/bin/.ssh2john.py-wrapped", line 104, in read_private_key
    data = base64.decodestring(data)
           ^^^^^^^^^^^^^^^^^^^
AttributeError: module 'base64' has no attribute 'decodestring'

```

### Notify maintainers

@offlinehacker 
@matthewbauer 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.1-zen1, NixOS, 23.11 (Tapir), 23.11.20231112.e44462d`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos-23.05, nixpkgs"`
 - channels(casual): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by CherryKitten**
## Description of changes

Nixpkgs is currently still on a 5 year old release, because upstream has not been tagging new stable releases for a very long time.
They do provide some form of rolling release in a seperate repository (https://github.com/openwall/john-packages/releases/tag/rolling-2404), so I have decided to use the same git commit hash as was used in the latest rolling release.

Some parts of the software have been broken on nixpkgs for a while. Updating to this rolling release will for example fix https://github.com/NixOS/nixpkgs/issues/267978, which has been open for almost a year now.

I've also added myself to the maintainer list.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by itslychee** [comment]
looks like there's a new release being planned here as well!

https://github.com/openwall/john/pull/5494

**Comment by litchipi** [comment]
When trying to use this version to crack a `yescrypt` hash, it tells me he cannot find any `john.conf` in the nix store `/share/john` directory.
`fopen: /nix/store/slkgvc2kkpiiz5pgyjjgzgpgqka70y2w-john-rolling-2404/share/john/john.conf: No such file or directory`

And when I setup a local `--config=john.conf` replacement, it seeks for a `dumb16.conf` which he doesn't find either.
I'm not really familiar with how john works internally, but it seams some files are missing from the install phases ?

**Comment by CherryKitten** [comment]
> looks like there's a new release being planned here as well!
> 
> [openwall/john#5494](https://github.com/openwall/john/pull/5494)

huh, created a day after my PR here, looks like the old issue being pushed up again might have set something in motion.

@litchipi i'll have a closer look soon, john is a pretty big and complex project so there is a good chance I might have missed parts in the update process that are only neccessary for use-cases I didn't test myself

**Comment by CherryKitten** [comment]
> When trying to use this version to crack a `yescrypt` hash, it tells me he cannot find any `john.conf` in the nix store `/share/john` directory. `fopen: /nix/store/slkgvc2kkpiiz5pgyjjgzgpgqka70y2w-john-rolling-2404/share/john/john.conf: No such file or directory`
> 
> And when I setup a local `--config=john.conf` replacement, it seeks for a `dumb16.conf` which he doesn't find either. I'm not really familiar with how john works internally, but it seams some files are missing from the install phases ?

this should be fixed now, this was covered by the steps in the `postPath` phase, which I had removed together with the patches, despite still being needed.

I added them back and tried it out, it no longer asks for the config file

**Comment by litchipi** [comment]
Ok, it fixed the config problem error, but I still have a `yescrypt` problem.
Could you test it on your machine ? Running john on this hash `root:$y$j9T$AjVXCCcjJ6jTodR8BwlPf.$4NeBwxOq4X0/0nCh3nrIBmwEEHJ6/kDU45031VFCWc2` (it comes from HackMyVM)

It seams that even with `libxcrypt` present (at least I see it listed in `ldd`), it doesn't detect that the system supports `yescrypt`.
Browsed a bit the sources, I thought it could come from https://github.com/openwall/john/blob/6be5461b67951290f707021b93b34ea4168fa9e2/src/c3_fmt.c#L341, OpenMP calling the `crypt_r` function, which apparently is at the collision between `unistd` and `libxcrypt` (as explained here: https://github.com/besser82/libxcrypt/blob/72f75aa370ae96ccd2cc44ea3cf4182d8679ffbe/lib/crypt-port.h#L53)
But maybe I got it all wrong, not really sure what I'm reading tbh.

Tell me if that's the for another issue, or if we can try to tackle this here before merging the update ?

**Comment by CherryKitten** [comment]
> Ok, it fixed the config problem error, but I still have a `yescrypt` problem. Could you test it on your machine ? Running john on this hash `root:$y$j9T$AjVXCCcjJ6jTodR8BwlPf.$4NeBwxOq4X0/0nCh3nrIBmwEEHJ6/kDU45031VFCWc2` (it comes from HackMyVM)
> 
> It seams that even with `libxcrypt` present (at least I see it listed in `ldd`), it doesn't detect that the system supports `yescrypt`. Browsed a bit the sources, I thought it could come from https://github.com/openwall/john/blob/6be5461b67951290f707021b93b34ea4168fa9e2/src/c3_fmt.c#L341, OpenMP calling the `crypt_r` function, which apparently is at the collision between `unistd` and `libxcrypt` (as explained here: https://github.com/besser82/libxcrypt/blob/72f75aa370ae96ccd2cc44ea3cf4182d8679ffbe/lib/crypt-port.h#L53) But maybe I got it all wrong, not really sure what I'm reading tbh.
> 
> Tell me if that's the for another issue, or if we can try to tackle this here before merging the update ?

yeah, i'm getting the same problem and also can't figure out a solution right now, sorry :/

Maybe since this was broken before already, it'd make sense to merge the update, since that fixes various other issues, and then keep focussing on the yescrypt problem seperately in a new issue

**Comment by JohnRTitor** [comment]
Result of `nixpkgs-review pr 318620` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>adenum</li>
    <li>john</li>
    <li>johnny</li>
    <li>wifite2</li>
    <li>wifite2.dist</li>
  </ul>
</details>

**Review by JohnRTitor — APPROVED** [review]
Builds and runs, lgtm

File: pkgs/tools/security/john/default.nix
Comment (JohnRTitor) [inline pkgs/tools/security/john/default.nix:6]: Unstable versioning should be formatted like: `1.9.0-jumbo-unstable-YYYY-MM-DD`

File: pkgs/tools/security/john/default.nix
Comment (JohnRTitor) [inline pkgs/tools/security/john/default.nix:10]:
```suggestion
    repo = "john";
```

**Review by JohnRTitor — COMMENTED** [review]
Could you address these before I merge?

File: pkgs/tools/security/john/default.nix
Comment (CherryKitten) [reply pkgs/tools/security/john/default.nix:6]: are you sure? "rolling-2024" is a version name upstream uses in their unstable packages, so i went with that over a custom nixpkgs specific version name

Comment (JohnRTitor) [reply pkgs/tools/security/john/default.nix:6]: According #234201, we should include the preceeding version if we are packaging something untagged. `rolling-2024` does not really a "version tag", is it?

Comment (CherryKitten) [reply pkgs/tools/security/john/default.nix:6]: it is a version tag upstream at https://github.com/openwall/john-packages/releases/tag/rolling-2404

i don't have a strong opinion on it either way, i can still change it when I'm able to tomorrow if you think it's really important

File: pkgs/tools/security/john/default.nix
Comment (JohnRTitor) [inline pkgs/tools/security/john/default.nix:11]: Then can we use 
```suggestion
    rev = "ref/tags/${version}";
```

File: pkgs/tools/security/john/default.nix
Comment (JohnRTitor) [reply pkgs/tools/security/john/default.nix:6]: Oh and looks like previous versions are also tagged https://github.com/openwall/john-packages/releases/tag/rolling-2310 https://github.com/openwall/john-packages/releases/tag/rolling-2304 

Then I am open to keeping the current version `rolling-2404`. And instead of the commit, the tag can be directly fetched I suppose? See my comment below

File: pkgs/tools/security/john/default.nix
Comment (JohnRTitor) [inline pkgs/tools/security/john/default.nix:12]: `hash` is preferred

```suggestion
    hash = "sha256-zvoN+8Sx6qpVg2JeRLOIH1ehfl3tFTv7r5wQZ44Qsbc=";
```

File: pkgs/tools/security/john/default.nix
Comment (CherryKitten) [reply pkgs/tools/security/john/default.nix:6]: that doesn't work, because the tag is in a seperate repo, upstream is a bit messy there..

they've tagged that version in their "john-packages" repo, which contains some binary builds and instructions for packaging and stuff, but it's not tagged in the actual "john" repo containing the code of the project, which is the one we fetch

that's why I've just used the same commit hash that's referenced in their package repo

Comment (CherryKitten) [reply pkgs/tools/security/john/default.nix:6]: to make this a bit clearer

the upstream organization has two repos

john
john-packages

john has the actual code and some old version tags, john-packages has some scripts for packaging 

on john-packages, there is an official version tag for rolling-2404, with packaging instructions that refer back to the john repo at a specific commit hash

that is why i used the version name from john-packages and directly fetch the corresponding code from john

Comment (JohnRTitor) [reply pkgs/tools/security/john/default.nix:6]: Generic update scripts/bots are gonna have a hard time dealing with that. Wish there was a update script. 

If you'd like to design an update script for it, you can fetch the tag from `john-packages` repo, and somehow get the commit hash from `Created-on_2024-04-03.txt` release file. Then substitute the commit hash in the package definition. 

I'd assume that would be a lot complex to design, I am not opposed to the idea of fetching the latest commit/making it stay on bleeding edge all the time. There is `unstableGitUpdater` for that too, so bots like r-ryantm can easily update it even in a maintainer crisis.

File: pkgs/tools/security/john/default.nix
Comment (JohnRTitor) [reply pkgs/tools/security/john/default.nix:11]: Not applicable

File: pkgs/tools/security/john/default.nix
Comment (CherryKitten) [reply pkgs/tools/security/john/default.nix:6]: This package has already not been updated in like 5 years, I'd rather just get any more recent version merged than bikeshedding about update-scripts to be honest..

I fixed the other comments you had, but I would prefer to just stay on the version and method we have and not do any more big changes, something like update scripts might be a thing for a future follow-up in my opinion

**Pull request merged**
